perf: various improvements to `fn decode_coefs`

daxtens opened the pull request:
This takes the existing PR #1405 and adds 3 changes on top to further improve the performance of `decode_coefs`. Each has detailed explanations in the respective  commit message, but to summarise:

 - statically determine the maximum result of `get_lo_ctx`, avoiding a bounds check before accessing an element of `lo_cdf`. Not a big win, but - like the change to get rid of the bounds check in `get_lo_ctx`, it's nice.
 
  - replace the bespoke `.fill(0)` implementation:
    * Previously, the `H` and `V` cases were able to be statically bounds checked, and that made them sufficiently simple that LLVM was able to recognise the loop pattern as a zeroing loop and replace it with a call to `memset` / `bzero`. 
    * In the `TwoD` case did the loop not get replaced, but instead split into a vectorised part and a non-vectorised part. The vectorised part did not bounds check, but the non-vectorised byte-by-byte part did bounds check on _every iteration_.
    * If we provide enough for the compiler to be able to statically determine that a bounds check is not needed in the `TwoD` case, we could avoid that check in each iteration. That's annoying but straightforward.
    * Once we do that, we're probably best served using `fill` (and therefore `memset`) explicitly. At least in glibc, this is an optimised and cpu-specific implementation. The vectorised loop generated by the compiler will not be this cpu-specific: it will target a generic machine of the architecture, which especially in the case of x86-64 is quite dated.
    
 - Unroll the `ctx_offsets` array. In `get_lo_ctx`. On the two x86_64 systems I've tested, the conditional moves to do `min(4, x)` and `min(4, y)` add a lot of latency and we cannot do the memory lookup into `ctx_offsets` until they've completed. We can just unroll the array though, by just duplicating the final element to reach 32 columns, and then duplicating the final row to reach 32 rows. This trades off data-cache efficiency for decreased latency on the memory lookup, and on both x86 systems I tested, the unrolled array is faster.
 
This is an improvement on my very dated x86_64 linux laptop (Intel Core i7-7560U, dual core with hyperthreading, running with all 4 threads):

 - `perf` shows a reduction in cycles for `decode_coefs` with the summer_nature_1080p.ivf file:

```
# Baseline  Delta Abs  Shared Object      Symbol                                                               >
# ........  .........  .................  .....................................................................>
#
     9.34%     -1.18%  dav1d              [.] rav1d::src::recon::decode_coefs
     9.87%     +0.20%  dav1d              [.] rav1d::src::decode::decode_b
     2.27%     +0.16%  dav1d              [.] rav1d::src::refmvs::add_spatial_candidate
     5.04%     +0.15%  dav1d              [.] rav1d::src::refmvs::rav1d_refmvs_find
     0.56%     +0.13%  dav1d              [.] dav1d_msac_decode_bool_equi_sse2
     0.10%     +0.11%  libc.so.6          [.] __memset_avx2_unaligned_erms
     2.11%     -0.10%  dav1d              [.] dav1d_load_tmvs_sse4.xloop
     3.26%     +0.10%  dav1d              [.] dav1d_msac_decode_symbol_adapt4_sse2
     1.36%     -0.10%  dav1d              [.] dav1d_load_tmvs_sse4.write_loop_entry
```

There's a big drop in decode_coefs time. There's an lot of random noise, an an increase in memset from glibc, as would be expected because we're now using it for the TwoD case.

 - hyperfine shows a roughly 0-1% overall improvement across the cases I've tested.

I also tested on a Windows system with an AMD Ryzen 5 2600 6-core processor. Because I got a lot of system noise if I ran with maximum threads, I ran with 6 threads (1 per core). In this configuration, hyperfine benchmarking suggests that:
 - the 8bit Chimera test case is about 0.7% faster,
 - the 10bit Chimera test case about 0.3% faster

(For completeness: I have a Mac but it's an M1 MacBook Air so it's a nightmare to benchmark on: the thermal limiting makes the multithreaded process slower each time it runs! I was not able to observe any difference in speed on any single-threaded benchmark. Maybe Apple Silicon is just better! 🤣 )

kkysen commented on src/tables.rs:
```diff
@@ -503,27 +504,405 @@ pub static DAV1D_YMODE_SIZE_CONTEXT: [u8; BlockSize::COUNT] = [
+// move sequences that have to happen in series before we can actually load the element
+// we want. These add latency. Simply unroll the array: duplicate everything from the 5th
+// row and 5th column out so that we get the same effect  as clamping without the latency.
+// Obviously we pay for it in cache efficiency and binary size, but the  tradeoff seems
```
```suggestion
// Obviously we pay for it in cache efficiency and binary size, but the tradeoff seems
```

kkysen commented on src/tables.rs:
```diff
@@ -503,27 +504,405 @@ pub static DAV1D_YMODE_SIZE_CONTEXT: [u8; BlockSize::COUNT] = [
+// `[0, 32)` to `[0, 5)`. That involves two `min`s, which LLVM makes into conditional
+// move sequences that have to happen in series before we can actually load the element
+// we want. These add latency. Simply unroll the array: duplicate everything from the 5th
+// row and 5th column out so that we get the same effect  as clamping without the latency.
```
```suggestion
// row and 5th column out so that we get the same effect as clamping without the latency.
```

kkysen commented on src/tables.rs:
```diff
@@ -503,27 +504,405 @@ pub static DAV1D_YMODE_SIZE_CONTEXT: [u8; BlockSize::COUNT] = [
-        [1, 6, 6, 21, 21],
-        [6, 6, 21, 21, 21],
-        [6, 21, 21, 21, 21],
-        [21, 21, 21, 21, 21],
```
For readability, can we do something like this instead?

```rust
/// Extend `a` from length `M` to length `N`, repeating the last element.
/// That is, `a[cmp::min(i, M - 1)] = extend_array(a)[cmp::min(i, N - 1)]`.
const fn extend_array<T: DefaultValue + Copy, const M: usize, const N: usize>(a: [T; M]) -> [T; N] {
    let mut b = [DefaultValue::DEFAULT; N];
    const_for!(i in 0..M => {
        b[i] = a[i];
    });
    const_for!(i in M..N => {
        b[i] = a[M - 1];
    });
    b
}

pub type LoCtxOffset = InRange<u8, 0, 21>;

pub static DAV1D_LO_CTX_OFFSETS: [[[LoCtxOffset; 32]; 32]; 3] = {
    const fn extend_offsets(a: [[u8; 5]; 5]) -> [[LoCtxOffset; 32]; 32] {
        extend_array([
            extend_array(LoCtxOffset::new_array(a[0])),
            extend_array(LoCtxOffset::new_array(a[1])),
            extend_array(LoCtxOffset::new_array(a[2])),
            extend_array(LoCtxOffset::new_array(a[3])),
            extend_array(LoCtxOffset::new_array(a[4])),
        ])
    }

    [
        extend_offsets([
            [0, 1, 6, 6, 21],
            [1, 6, 6, 21, 21],
            [6, 6, 21, 21, 21],
            [6, 21, 21, 21, 21],
            [21, 21, 21, 21, 21],
        ]),
        extend_offsets([
            [0, 16, 6, 6, 21],
            [16, 16, 6, 21, 21],
            [16, 16, 21, 21, 21],
            [16, 16, 21, 21, 21],
            [16, 16, 21, 21, 21],
        ]),
        extend_offsets([
            [0, 11, 11, 11, 11],
            [11, 11, 11, 11, 11],
            [6, 6, 21, 21, 21],
            [6, 21, 21, 21, 21],
            [21, 21, 21, 21, 21],
        ]),
    ]
};
```

And then this added to `InRange`:

```rust
        impl<const N: usize, const MIN: u128, const MAX: u128> DefaultValue for [InRange<$T, MIN, MAX>; N] {
            const DEFAULT: Self = [DefaultValue::DEFAULT; N];
        }
```

I can add that as a commit if that looks good.

daxtens commented:
Yeah, that macro and addition to inrange is definitely an improvement. I am
happy for you to commit that (and the comment wording changes if you want
to)

On Sat, 7 Jun 2025 at 7:41 am, Khyber Sen ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/tables.rs
> <https://github.com/memorysafety/rav1d/pull/1425#discussion_r2132912566>:
>
> > @@ -503,27 +504,405 @@ pub static DAV1D_YMODE_SIZE_CONTEXT: [u8; BlockSize::COUNT] = [
>      3, 3, 3, 3, 3, 2, 3, 3, 2, 1, 2, 2, 2, 1, 0, 1, 1, 1, 0, 0, 0, 0,
>  ];
>
> -pub static DAV1D_LO_CTX_OFFSETS: [[[u8; 5]; 5]; 3] = [
> +pub type LoCtxOffset = InRange<u8, 0, 21>;
> +// The core of each of these 3 tables is a 5x5 table of offsets. However, to use that
> +// 5x5 table in `get_lo_ctx`, we first have to clamp the `x` and `y` values from
> +// `[0, 32)` to `[0, 5)`. That involves two `min`s, which LLVM makes into conditional
> +// move sequences that have to happen in series before we can actually load the element
> +// we want. These add latency. Simply unroll the array: duplicate everything from the 5th
> +// row and 5th column out so that we get the same effect  as clamping without the latency.
> +// Obviously we pay for it in cache efficiency and binary size, but the  tradeoff seems
>
> ⬇️ Suggested change
>
> -// Obviously we pay for it in cache efficiency and binary size, but the  tradeoff seems
> +// Obviously we pay for it in cache efficiency and binary size, but the tradeoff seems
>
> ------------------------------
>
> In src/tables.rs
> <https://github.com/memorysafety/rav1d/pull/1425#discussion_r2132912709>:
>
> > @@ -503,27 +504,405 @@ pub static DAV1D_YMODE_SIZE_CONTEXT: [u8; BlockSize::COUNT] = [
>      3, 3, 3, 3, 3, 2, 3, 3, 2, 1, 2, 2, 2, 1, 0, 1, 1, 1, 0, 0, 0, 0,
>  ];
>
> -pub static DAV1D_LO_CTX_OFFSETS: [[[u8; 5]; 5]; 3] = [
> +pub type LoCtxOffset = InRange<u8, 0, 21>;
> +// The core of each of these 3 tables is a 5x5 table of offsets. However, to use that
> +// 5x5 table in `get_lo_ctx`, we first have to clamp the `x` and `y` values from
> +// `[0, 32)` to `[0, 5)`. That involves two `min`s, which LLVM makes into conditional
> +// move sequences that have to happen in series before we can actually load the element
> +// we want. These add latency. Simply unroll the array: duplicate everything from the 5th
> +// row and 5th column out so that we get the same effect  as clamping without the latency.
>
> ⬇️ Suggested change
>
> -// row and 5th column out so that we get the same effect  as clamping without the latency.
> +// row and 5th column out so that we get the same effect as clamping without the latency.
>
> ------------------------------
>
> In src/tables.rs
> <https://github.com/memorysafety/rav1d/pull/1425#discussion_r2132938369>:
>
> > +pub static DAV1D_LO_CTX_OFFSETS: [[[LoCtxOffset; 32]; 32]; 3] = [
>      [
> -        [0, 1, 6, 6, 21],
> -        [1, 6, 6, 21, 21],
> -        [6, 6, 21, 21, 21],
> -        [6, 21, 21, 21, 21],
> -        [21, 21, 21, 21, 21],
>
> For readability, can we do something like this instead?
>
> /// Extend `a` from length `M` to length `N`, repeating the last element./// That is, `a[cmp::min(i, M - 1)] = extend_array(a)[cmp::min(i, N - 1)]`.const fn extend_array<T: DefaultValue + Copy, const M: usize, const N: usize>(a: [T; M]) -> [T; N] {
>     let mut b = [DefaultValue::DEFAULT; N];
>     const_for!(i in 0..M => {
>         b[i] = a[i];
>     });
>     const_for!(i in M..N => {
>         b[i] = a[M - 1];
>     });
>     b}
> pub type LoCtxOffset = InRange<u8, 0, 21>;
> pub static DAV1D_LO_CTX_OFFSETS: [[[LoCtxOffset; 32]; 32]; 3] = {
>     const fn extend_offsets(a: [[u8; 5]; 5]) -> [[LoCtxOffset; 32]; 32] {
>         extend_array([
>             extend_array(LoCtxOffset::new_array(a[0])),
>             extend_array(LoCtxOffset::new_array(a[1])),
>             extend_array(LoCtxOffset::new_array(a[2])),
>             extend_array(LoCtxOffset::new_array(a[3])),
>             extend_array(LoCtxOffset::new_array(a[4])),
>         ])
>     }
>
>     [
>         extend_offsets([
>             [0, 1, 6, 6, 21],
>             [1, 6, 6, 21, 21],
>             [6, 6, 21, 21, 21],
>             [6, 21, 21, 21, 21],
>             [21, 21, 21, 21, 21],
>         ]),
>         extend_offsets([
>             [0, 16, 6, 6, 21],
>             [16, 16, 6, 21, 21],
>             [16, 16, 21, 21, 21],
>             [16, 16, 21, 21, 21],
>             [16, 16, 21, 21, 21],
>         ]),
>         extend_offsets([
>             [0, 11, 11, 11, 11],
>             [11, 11, 11, 11, 11],
>             [6, 6, 21, 21, 21],
>             [6, 21, 21, 21, 21],
>             [21, 21, 21, 21, 21],
>         ]),
>     ]};
>
> And then this added to InRange:
>
>         impl<const N: usize, const MIN: u128, const MAX: u128> DefaultValue for [InRange<$T, MIN, MAX>; N] {
>             const DEFAULT: Self = [DefaultValue::DEFAULT; N];
>         }
>
> I can add that as a commit if that looks good.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/memorysafety/rav1d/pull/1425#pullrequestreview-2906152133>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAZHSBL7QPYUCLWWEYTOU5T3CIDIPAVCNFSM6AAAAAB6YIOQKCVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDSMBWGE2TEMJTGM>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

kkysen commented:
> Yeah, that macro and addition to inrange is definitely an improvement. I am happy for you to commit that (and the comment wording changes if you want to)

Okay, I rebased and added that commit.  I left the other little wording changes for now, as I'll let you rebase those (I'll let you manage those commits; don't wanna conflict accidentally).

daxtens commented:
Ok, I've applied those two suggested changes, thank you. 

The branch currently sits cleanly on top of main, so I don't think there's any more rebasing required at this stage. Let me know if there's anything else you'd like on this PR, in the mean time I will be working on some of the stuff around `RefMvsCandidate`!

kkysen commented:
By the way, for commits like 532d182, can you give it a more descriptive commit message?  Or just rebase if that makes more sense.

kkysen commented on src/recon.rs:
```diff
@@ -817,33 +840,75 @@ fn decode_coefs<BD: BitDepth>(
+                //  `32 * (32 + 2) = 1088`
+                // Therefore, no bounds check should be required to slice `levels`.
+                //
+                // However, as of nightly 2025-05-01, rustc had a lot of trouble
+                // proving that to itself. Here we give it a hand. First, we
```
Even with extra `assert!`s, it can't do this?  That's surprising.  Usually LLVM is better than that.

daxtens replied:
Oh sorry, I totally misunderstood the context of your comment. With the rearranged calculation , the asserts on `term1, `term2, and the final bounds check before `fill` are all elided. The only thing that I can't get regardless of asserts is the addition vs bitwise or.

daxtens commented on src/recon.rs:
```diff
@@ -505,22 +505,45 @@ fn get_lo_ctx(
+    // value in the underlying table is 21, and we guarantee that with types.
+    //
+    // So at this point, `offset <= 36`. However, it seems the compiler needs
+    // some help to unify the offset maxima from the branches. This seems to
```
With the assert, which is elided, it does OK. I've made the elision clearer in the comment.

daxtens commented on src/recon.rs:
```diff
@@ -472,17 +473,21 @@ fn get_dc_sign_ctx(tx: TxfmSize, a: &[u8], l: &[u8]) -> c_uint {
+// This is used for lo_ctx, which is used for a lookup into lo_cdf. lo_cdf corresponds
+// roughly to Coeff_Base_Cdf in the spec, rav1d/dav1d uses a  maximum value is 40 (for
+// an array size of 41), whereas the spec has the array being 42 elements.
+type LoCdfIndex = InRange<usize, 0, 40>;
```
It's not stored anywhere - it leaves the function and is immediately used as an index into lo_cdf, see e.g. line 1019: https://github.com/memorysafety/rav1d/pull/1425/files#diff-12d3a471419f71ca340bb39fc648e70a9b3dc9b3aef6bf125b894d8abb95c60aR1019 . Using `usize` instead of `u8` just saves us a cast at the index sites.

kkysen replied:
Ah, okay.  I think it would be fine then, as the range is very clear now if we ever need to change this in the future in case it gets stored or something.

daxtens commented on src/recon.rs:
```diff
@@ -472,17 +473,21 @@ fn get_dc_sign_ctx(tx: TxfmSize, a: &[u8], l: &[u8]) -> c_uint {
     levels: &[u8],
     tx_class: TxClass,
     hi_mag: &mut u32,
-    ctx_offsets: Option<&[[u8; 5]; 5]>,
+    ctx_offsets: Option<&[[LoCtxOffset; 32]; 32]>,
     x: u8,
     y: u8,
-    stride: u8,
-) -> u8 {
-    let stride = stride as usize;
+    stride: usize,
```
Not strictly. 

It saves us doing the `let stride = stride as usize` in the body of the function. 

Because of type inference, it doesn't require any changes to the caller code. `stride` is defined in the caller at line 829 https://github.com/memorysafety/rav1d/pull/1425/files#diff-12d3a471419f71ca340bb39fc648e70a9b3dc9b3aef6bf125b894d8abb95c60aR829 and the definition doesn't have a type, so this change causes Rust to fill in the type of `stride` in the caller as `usize` rather than `u8`.

I think it's a slight readability improvement not to take a parameter and then immediately cast it to a different numeric type, but I'm not much fussed and I can remove it if you'd prefer.

kkysen replied:
Hmm.  I'd been preferring the smaller type size for clarity, as it strictly tells you more information, and when it gets stored, it's more efficient.  For an `InRange` type, it's a bit different because the range is explicit.

daxtens replied:
No worries. Unpicked this in the latest force-push.

daxtens commented:
Thanks for those - I have squashed them into the relevant commits and rebased on main.

And don't worry about being nitpicky - you have to have an eye to actually maintaining this in the long term, while I just look for interesting performance puzzles. And your feedback is very actionable and specific, which makes applying it very straightforward.

kkysen commented on src/recon.rs:
```diff
@@ -505,22 +511,43 @@ fn get_lo_ctx(
+            ((mag + 64) >> 7) as u8 as usize
+        } as usize;
+
+    // `36 + 4 == 40`, so the unwrap should be elided
+    LoCdfIndex::new(lo_ctx).unwrap()
```
```suggestion
    // `36 + 4 == 40`.
    LoCdfIndex::new(lo_ctx).unwrap() // Elided
```

kkysen reviewed (approved):
A couple of little comments, but I think this looks good now.  Thanks!

kkysen reviewed (approved):
> Ok, I've pushed with that final comment tweak, and with unpicking the needless u8->usize change.

Sorry, I meant to merge this earlier, but #1430 just merged, which causes a lot of merge conflicts.  If you can fix them, I'll merge this, or I can try to fix them (they're a pain).
